### PR TITLE
Optimize j/l/Class.isAssignableFrom on X86

### PR DESCRIPTION
**Pull request by 0dvictor**
java/lang/Class.isAssignableFrom() is transformed into a JIT Helper
so that JNI call overhead is avoided.

Signed-off-by: Victor Ding <dvictor@ca.ibm.com>

### CODE REVIEW

**Comment by 0dvictor** [comment]
This optimization in theory applies to all platforms; however, some platforms may need extra work in their CodeGen to take advantage of it. For example, Z has its own optimizations in its CodeGen to fast-path certain cases. Should this change be applied to Z blindly, its optimizations would be bypassed and hence the performance may be negatively impacted.
@r30shah FYI.

File: runtime/compiler/x/codegen/J9CodeGenerator.cpp [hunk 1]
```diff
@@ -396,6 +396,13 @@ J9::X86::CodeGenerator::suppressInliningOfRecognizedMethod(TR::RecognizedMethod
       }
    }
 
+bool
+J9::X86::CodeGenerator::supportsInliningOfIsAssignableFrom()
+   {
+   static auto disableInliningOfIsAssignableFrom = (bool)feGetEnv("TR_DisableInliningOfIsAssignableFrom");
```
[the remark this answers is not in the document]
Comment (0dvictor) [reply runtime/compiler/x/codegen/J9CodeGenerator.cpp]: I see your point to make LHS `const`, the code has been updated.

**Comment by cathyzhyi** [comment]
LGTM

File: runtime/oti/jitprotos.h [hunk 2]
```diff
@@ -163,8 +163,10 @@ void J9FASTCALL fast_jitCollapseJNIReferenceFrame(J9VMThread *currentThread);
 #if defined(J9VM_ARCH_X86) || defined(J9VM_ARCH_S390)
 /* TODO Will be cleaned once all platforms adopt the correct parameter order */
 UDATA J9FASTCALL fast_jitInstanceOf(J9VMThread *currentThread, j9object_t object, J9Class *castClass);
+UDATA J9FASTCALL fast_jitCheckAssignable(J9VMThread *currentThread, J9Class *clazz, J9Class *castClass);
 #else /* J9VM_ARCH_X86 || J9VM_ARCH_S390*/
 UDATA J9FASTCALL fast_jitInstanceOf(J9VMThread *currentThread, J9Class *castClass, j9object_t object);
+UDATA J9FASTCALL fast_jitCheckAssignable(J9VMThread *currentThread, J9Class *castClass, J9Class *clazz)
```
Comment (fjeremic) [inline runtime/oti/jitprotos.h]: I think we may be missing a semicolon here?

Comment (0dvictor) [reply runtime/oti/jitprotos.h]: Good catch! It is fixed.

File: runtime/codert_vm/cnathelp.cpp [hunk 3]
```diff
@@ -2517,6 +2517,18 @@ fast_jitInstanceOf(J9VMThread *currentThread, J9Class *castClass, j9object_t obj
 	return isInstance;
 }
 
+UDATA J9FASTCALL
+#if defined(J9VM_ARCH_X86) || defined(J9VM_ARCH_S390)
+/* TODO Will be cleaned once all platforms adopt the correct parameter order */
+fast_jitCheckAssignable(J9VMThread *currentThread, J9Class *clazz, J9Class *castClass)
+#else /* J9VM_ARCH_X86 || J9VM_ARCH_S390*/
+fast_jitCheckAssignable(J9VMThread *currentThread, J9Class *castClass, J9Class *clazz)
+#endif /* J9VM_ARCH_X86 || J9VM_ARCH_S390*/
```
Comment (fjeremic) [inline runtime/codert_vm/cnathelp.cpp]: Are there some braces missing around this area? Not really sure what is going on but looking at the surrounding code this appears to be incorrect at first glance. Perhaps I'm missing something.

Comment (0dvictor) [reply runtime/codert_vm/cnathelp.cpp]: I don't think extra braces are need here. It looks very similar to the method above `fast_jitInstanceOf` and compiles.

**Comment by 0dvictor** [comment]
Windows failure due to #2129

**Pull request merged**